controllers/dockermachine_controller.go

```diff
@@ -47,17 +68,257 @@ func (r *DockerMachineReconciler) Reconcile(req ctrl.Request) (ctrl.Result, erro
+		log.Info("did not find matching machine reference")
+		return ctrl.Result{}, nil
+	}
+
+	// Get the cluster api machine
+	machine := &capiv1alpha2.Machine{}
+	machineKey := client.ObjectKey{
+		Namespace: req.Namespace,
+		Name:      machineRef.Name,
+	}
+
+	if err := r.Get(ctx, machineKey, machine); err != nil {
+		log.Error(err, "failed to get machine")
+		return ctrl.Result{}, err
+	}
+
+	if machine.Labels[capiv1alpha2.MachineClusterLabelName] == "" {
+		return ctrl.Result{}, errors.New("machine has no associated cluster")
+	}
+
+	// Get the cluster
+	cluster := &capiv1alpha2.Cluster{}
+	clusterKey := client.ObjectKey{
+		Namespace: req.Namespace,
+		Name:      machine.Labels[capiv1alpha2.MachineClusterLabelName],
+	}
+	if err := r.Get(ctx, clusterKey, cluster); err != nil {
+		log.Error(err, "failed to get cluster")
+		return ctrl.Result{}, err
+	}
+
+	// creating loadbalancer
+	log.Info("Reconciling cluster")
+	elb, err := getExternalLoadBalancerNode(cluster.Name, log)
+	if err != nil {
+		log.Error(err, "Error getting external load balancer node")
+		return ctrl.Result{}, err
+	}
+	if elb == nil {
+		log.Info("External Load Balancer does not exist. Create ELB.")
+		_, err = actions.SetUpLoadBalancer(cluster.Name)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	// create docker node
+	if dockerMachine.Spec.ProviderID != nil {
+		return ctrl.Result{}, nil
+	}
+
+	result, err := r.create(ctx, cluster, machine, dockerMachine)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if err := r.Client.Update(ctx, dockerMachine); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// TODO should the cluster be deleted?
+
+	return result, nil
+}
+
+func getRole(machine *capiv1alpha2.Machine) string {
+	// Figure out what kind of node we're making
+	labels := machine.GetLabels()
+	setValue, ok := labels["set"]
+	if !ok {
+		setValue = constants.WorkerNodeRoleValue
+	}
+	return setValue
+}
+
+func (r *DockerMachineReconciler) create(
+	ctx context.Context,
+	c *capiv1alpha2.Cluster,
+	machine *capiv1alpha2.Machine,
+	dockerMachine *infrastructurev1alpha2.DockerMachine,
+) (ctrl.Result, error) {
+	r.Log.Info("Creating a machine for cluster", "cluster-name", c.Name)
+	clusterExists, err := cluster.IsKnown(c.Name)
+	if err != nil {
+		r.Log.Error(err, "Error finding cluster-name", "cluster", c.Name)
+		return ctrl.Result{}, err
+	}
+	// If there's no cluster, requeue the request until there is one
+	if !clusterExists {
+		r.Log.Info("There is no cluster yet, waiting for a cluster before creating machines")
+		return ctrl.Result{
+			RequeueAfter: time.Second * 30,
+		}, nil
+	}
+	controlPlanes, err := actions.ListControlPlanes(c.Name)
+	if err != nil {
+		r.Log.Error(err, "Error listing control planes")
+		return ctrl.Result{}, err
+	}
+	r.Log.Info("Is there a cluster?", "cluster-exists", clusterExists)
+	setValue := getRole(machine)
+	r.Log.Info("This node has a role", "role", setValue)
+	if setValue == clusterAPIControlPlaneSetLabel {
```

Comment:
Why does an infra provider care about the role of the Machine (control plane or not)?

Reply:
The infra provider will need to know in order to apply the right security groups and the right IAM role for the case of AWS for example.

It will also need to know if it is modifying load balancer membership (such as AWS does as well)

Reply:
Makes sense. I was really hoping the infra providers could ignore the type of Machine, but I guess not 😄.

Reply:
I thiiiiiIIIIIIiiiink this could get simplified since in all cases it should run whatever comes back from the bootstrap provider. but the bootstrap provider will have to know what to generate

Reply:
bah but it does have to register with the LB and upload secret and stuff like that. Maybe not.